### PR TITLE
[Store] add TP chunk metadata put APIs 

### PR DESCRIPTION
**Pull request by zxpdemonio**
## Description                                                                                                                                                                     
                                                                                                                                                                                     
  This PR extracts the PR1-scoped TP metadata write path from the larger TP/CCRP work.                                                                                               
                                                                                                                                                                                     
  It adds chunk-oriented TP put APIs in the Python store bindings so callers that already own the sharding plan can write shard payloads together with TP metadata, without depending
   on later reconstruction/CCRP changes.                                                                                                                                             
                                                                                                                                                                                     
  Changes in this PR include:                                                                                                                                                        
  - adding `GlobalMetadata` and `ChunkMetadata` for TP shard metadata                                                                                                                
  - adding TP chunk put APIs:                                                                                                                                                        
    - `put_tensor_chunk_with_tp`                                                                                                                                                     
    - `batch_put_tensor_chunk_with_tp`                                                                                                                                               
    - `put_tensor_chunk_with_tp_from`                                                                                                                                                
    - `batch_put_tensor_chunk_with_tp_from`                                                                                                                                          
  - routing existing TP put paths through the same metadata-aware chunk put flow                                                                                                     
  - adding PR1-scoped tests for chunk put and zero-copy chunk put                                                                                                                    
  - updating Python API docs for the new APIs                                                                                                                                        
                      
## Motivation                                                                                                                                                               
This PR introduces tensor chunk write APIs that allow distributed processes to write their own tensor shards directly, instead of requiring the full tensor to be materialized on a single process first. The main motivation is to let each rank write its local tensor chunk concurrently, which avoids the extra all-gather cost and reduces both communication overhead and temporary memory pressure in distributed workloads. This is especially useful when each process already owns its local shard and only needs to persist it together with enough metadata for later reconstruction.                                                                                                                                     
                                                                                                                                                                                     
  It also prepares the storage interface for future heterogeneous tensor-parallel (TP) read/write scenarios. By making shard-level writes and metadata handling explicit now, we can support more flexible TP layouts and heterogeneous execution environments in follow-up work, where different devices or ranks may not share the same data placement or reconstruction path.
                                                                                                                                                                                     
  ## Module                                                                                                                                                                          
                                                                                                                                                                                     
  - [ ] Transfer Engine (`mooncake-transfer-engine`)                                                                                                                                 
  - [ ] Mooncake Store (`mooncake-store`)                                                                                                                                            
  - [ ] Mooncake EP (`mooncake-ep`)                                                                                                                                                  
  - [x] Integration (`mooncake-integration`)                                                                                                                                         
  - [ ] P2P Store (`mooncake-p2p-store`)                                                                                                                                             
  - [ ] Python Wheel (`mooncake-wheel`)                                                                                                                                              
  - [ ] PyTorch Backend (`mooncake-pg`)                                                                                                                                              
  - [ ] Mooncake RL (`mooncake-rl`)                                                                                                                                                  
  - [ ] CI/CD                                                                                                                                                                        
  - [x] Docs                                                                                                                                                                         
  - [ ] Other                                                                                                                                                                        
                                                                                                                                                                                     
  ## Type of Change                                                                                                                                                                  
                                                                                                                                                                                     
  - [ ] Bug fix                                                                                                                                                                      
  - [x] New feature                                                                                                                                                                  
  - [ ] Refactor                                                                                                                                                                     
  - [ ] Breaking change                                                                                                                                                              
  - [x] Documentation update                                                                                                                                                         
  - [ ] Other                                                                                                                                                                        
                                                                                                                                                                                     
  ## How Has This Been Tested?                                                                                                                                                       
                                                                                                                                                                                     
  - Performed self-review for:                                                                                                                                                       
    - maintainability / readability / duplicate logic                                                                                                                                
    - stability and memory-safety risks                                                                                                                                              
    - security concerns                                                                                                                                                              
    - static performance issues                                                                                                                                                      
    - doc/code consistency                                                                                                                                                           
  - Ran:                                                                                                                                                                             
    - `python -m py_compile scripts/test_tensor_chunk_api.py`                                                                                                                        
  - Added PR1-scoped tests in `scripts/test_tensor_chunk_api.py` covering:                                                                                                           
    - single chunk put                                                                                                                                                               
    - batch chunk put                                                                                                                                                                
    - zero-copy chunk put                                                                                                                                                            
    - metadata existence/content checks                                                                                                                                              
    - invalid `split_dim` negative case                                                                                                                                              
                                                                                                                                                                                     
  Note:                                                                                                                                                                              
  - full integration test execution was not run here because the current test script connects to a shared store and calls `remove_all()`.                                            
                                                                                                                                                                                     
  ## Checklist                                                                                                                                                                       
                                                                                                                                                                                     
  - [x] I have performed a self-review of my own code.                                                                                                                               
  - [x] I have formatted my own code using `./scripts/code_format.sh` before submitting.                                                                                             
  - [x] I have updated the documentation.                                                                                                                                            
  - [x] I have added tests to prove my changes are effective.

### CODE REVIEW

**Comment by zxpdemonio** [comment]
This pr is split from https://github.com/kvcache-ai/Mooncake/pull/1389

File: mooncake-integration/integration_utils.h [hunk 1]
```diff
@@ -96,10 +97,68 @@ inline TensorDtype get_tensor_dtype(py::object dtype_obj) {
     return TensorDtype::UNKNOWN;
 }
 
+inline size_t get_element_size(int32_t dtype) {
```
Comment (stmatengss) [inline mooncake-integration/integration_utils.h:100]: Why not directly use `sizeof`?

**Comment by zxpdemonio** [comment]
This PR introduces tensor chunk write APIs that allow distributed processes to write their own tensor shards directly, instead of requiring the full tensor to be materialized on a single process first. The main motivation is to let each rank write its local tensor chunk concurrently, which avoids the extra all-gather cost and reduces both communication overhead and temporary memory pressure in distributed workloads. This is especially useful when each process already owns its local shard and only needs to persist it together with enough metadata for later reconstruction.                                                                                                                                     
                                                                                                                                                                                     
  It also prepares the storage interface for future heterogeneous tensor-parallel (TP) read/write scenarios. By making shard-level writes and metadata handling explicit now, we can support more flexible TP layouts and heterogeneous execution environments in follow-up work, where different devices or ranks may not share the same data placement or reconstruction path.

File: mooncake-integration/integration_utils.h (hunk 1 above)
Comment (zxpdemonio) [reply mooncake-integration/integration_utils.h:100]: This follows the existing put_tensor serialization path.
In this code path we only have the runtime TensorDtype value from TensorMetadata, not a concrete compile-time C++ type, so sizeof(...) is not directly applicable here.
We therefore keep a runtime dtype-to-element-size mapping, consistent with the current put_tensor / put_tensor_from logic.

**Comment by zxpdemonio** [comment]
Performence test with different sizes：

  256MB                                                                                                                                                                              
                                                                                                                                                                                     
  - put: 39.14 ms                                                                                                                                                                    
  - put_from: 34.57 ms                                                                                                                                                               
  - chunk_put: 42.93 ms                                                                                                                                                              
  - chunk_put_from: 34.49 ms                                                                                                                                                         
                                                                                                                                                                                     
  512MB                                                                                                                                                                              
                                                                                                                                                                                     
  - put: 85.75 ms                                                                                                                                                                    
  - put_from: 72.54 ms                                                                                                                                                               
  - chunk_put: 94.01 ms                                                                                                                                                              
  - chunk_put_from: 74.27 ms                                                                                                                                                         
                                                                                                                                                                                     
  1024MB                                                                                                                                                                             
                                                                                                                                                                                     
  - put: 162.71 ms                                                                                                                                                                   
  - put_from: 130.16 ms                                                                                                                                                              
  - chunk_put: 158.90 ms                                                                                                                                                             
  - chunk_put_from: 134.58 ms 

And I profiled the overhead of metadata put：just 0.6ms。

We have not tested the performance of multiple threads concurrently writing their own chunks versus all-gather + put yet, because even when multiple threads write in parallel today, the underlying path is still effectively serialized. As a result, the expected advantage cannot be observed yet. We need to fix that issue first before the benefit can be reflected in benchmark results. For now, these performance tests show that there is no obvious performance regression.

**Comment by zxpdemonio** [comment]
We decided not to redesign the underlying metadata layer for TP chunk metadata at this stage.                                                                                      
 The main reason is that this requirement is currently very specific to tensor semantics, especially TP weight storage, rather than a general object-store requirement. The current global metadata + chunk metadata model is mainly used to describe tensor layout information such as full shape, split dimension, TP size, and per-chunk range. Other important workloads, such as KV cache, do not currently need the same abstraction. Because of that, pushing this logic down into a generic metadata mechanism would introduce broader system complexity for a use case that is still narrow in scope.
  We also do not think this is the highest-leverage optimization for the current performance issue. Based on our profiling so far, the main cost is still in the large shard data write path itself, not in the small metadata objects. Redesigning metadata storage would likely add significant implementation and maintenance complexity, while providing limited near-term performance benefit.
  For now, we believe it is better to keep TP-specific metadata in the tensor integration layer, where it remains clearly scoped to tensor/weight use cases. If similar requirements later appear in other workloads, or if we need a more general typed-object abstraction in the future, we can revisit a broader metadata redesign then.
